### PR TITLE
Specs

### PR DESCRIPTION
**Pull request by dareddov**

### CODE REVIEW

File: spec/graphql/services/order_by_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        third_product = create(:product, product_name: 'Product 3')
+
+        results = Services::OrderBy.call(Product, Product.all, 'product_name')
+
+        expect(results).to match_array([first_product, second_product, third_product])
+      end
+    end
+
+    context 'when provided arguments have orderBy' do
+      it 'return a sorted by product_name list of 3 projects' do
```
Comment (stevo) [inline spec/graphql/services/order_by_spec.rb]: descending

File: spec/graphql/services/order_by_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        first_product = create(:product, product_name: 'Product 1')
+        second_product = create(:product, product_name: 'Product 2')
+        third_product = create(:product, product_name: 'Product 3')
+
+        results = Services::OrderBy.call(Product, Product.all, '-product_name')
+
+        expect(results).to match_array([third_product, second_product, first_product])
```
Comment (stevo) [inline spec/graphql/services/order_by_spec.rb]: it does not assert order

File: spec/graphql/services/order_by_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      end
+    end
+
```
Comment (stevo) [inline spec/graphql/services/order_by_spec.rb]: ✂️

File: spec/graphql/functions/update_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          'product' =>
+            {
+              'id' => product.id,
+              'product_name' => 'New product name'
+            }
+        }
+
+        updated_product = Functions::Update.new(Product).call(nil, params, nil)
+
+        expect(updated_product.product_name).to eq 'New product name'
```
Comment (stevo) [inline spec/graphql/functions/update_spec.rb]: `product.reload`

File: spec/graphql/functions/update_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
+
+describe Functions::Update do
+
+  describe '.call' do
```
Comment (stevo) [inline spec/graphql/functions/update_spec.rb]: `#call`

File: spec/graphql/functions/find_by_id_spec.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      it 'return error record not found' do
+        params = { }
+
+        expect {
+          Functions::FindById.new(Product).call(nil, params, nil)
+        }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+  end
+end
```
Comment (stevo) [inline spec/graphql/functions/find_by_id_spec.rb]: line endings

File: spec/graphql/functions/find_by_id_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+
+describe Functions::FindById do
+
+  describe '.call' do
+    context 'when provided id of exists product' do
+      it 'return product' do
```
Comment (stevo) [inline spec/graphql/functions/find_by_id_spec.rb]: `when provided id of exists product` -> `when id of existing product is provided`
`returns`

File: spec/graphql/functions/find_by_id_spec.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        product = create(:product)
+        other_product = create(:product)
+        params = { id: product.id }
+
+        found_product = Functions::FindById.new(Product).call(nil, params, nil)
+
+        expect(found_product).to eq product
+        expect(found_product).not_to eq other_product
```
Comment (stevo) [inline spec/graphql/functions/find_by_id_spec.rb]: obsolete

File: spec/graphql/functions/find_by_id_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+describe Functions::FindById do
+
+  describe '.call' do
+    context 'when provided id of exists product' do
+      it 'return product' do
+        product = create(:product)
+        other_product = create(:product)
+        params = { id: product.id }
+
+        found_product = Functions::FindById.new(Product).call(nil, params, nil)
```
Comment (stevo) [inline spec/graphql/functions/find_by_id_spec.rb]: `found_product` -> `result`

File: spec/graphql/functions/find_all_spec.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
+require 'rails_helper'
+
+describe Functions::FindAll do
+
+  describe '.call' do
+    context 'when provided arguments are blank' do
+      it 'return a list of products' do
```
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: all

File: spec/graphql/functions/find_all_spec.rb [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        products = create_list(:product, 5)
```
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: 2

File: spec/graphql/functions/find_all_spec.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        params = { }
+
+        found_products = Functions::FindAll.new(Product).call(nil, params, nil)
+
+        expect(found_products).to eq products
+      end
+    end
+
+    context 'when provided arguments have offset argument' do
```
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: `when offset is provided`

File: spec/graphql/functions/find_all_spec.rb (hunk 10 above)
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: records

File: spec/graphql/functions/find_all_spec.rb [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      it 'return a list of 3 products' do
```
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: `returns all records starting from offset provided`

File: spec/graphql/functions/find_all_spec.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        products = create_list(:product, 5)
```
Comment (stevo) [inline spec/graphql/functions/find_all_spec.rb]: `_,  _,  product_3,  product_4 = create_list(:product, 4)`

File: spec/graphql/functions/delete_spec.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+  context 'when provided product exists' do
+    describe '.call' do
+      it 'destroys a product' do
+        product = create(:product)
+        params = { id: product.id }
+
+        product = Functions::Delete.new(Product).call(nil, params, nil)
+
+        expect(product.persisted?).to be_falsey
```
Comment (stevo) [inline spec/graphql/functions/delete_spec.rb]: `expect(Product.where(id: product.id)).to not_exist`

File: spec/graphql/functions/delete_spec.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+      end
+    end
+  end
+
+  context 'when provided product does not exists' do
```
Comment (stevo) [inline spec/graphql/functions/delete_spec.rb]: `when record with id provided does not exist`

File: spec/graphql/functions/create_spec.rb [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        product = Functions::Create.new(Product).call(nil, params, nil)
+
+        expect(product.persisted?).to be_truthy
+        expect(product).to have_attributes(product_name: 'Ergonomic Lamp')
+        expect(product.supplier).to have_attributes(name: 'Supplier #2')
+      end
+    end
+
+    context 'when blank parameters are provided' do
+      it 'not creates a product' do
```
Comment (stevo) [inline spec/graphql/functions/create_spec.rb]: `does not`

File: spec/graphql/functions/create_spec.rb [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        expect(product.supplier).to have_attributes(
+          name: 'Supplier #1',
+          webpage: 'http://example.com',
+          notes: 'Something note'
+        )
+
+      end
+    end
+
+    context 'when only require parameters are provided' do
```
Comment (stevo) [inline spec/graphql/functions/create_spec.rb]: `when valid parameters are provided`

File: spec/graphql/functions/create_spec.rb [hunk 19]
```diff
@@ -0,0 +1,80 @@
+require 'rails_helper'
+
+describe Functions::Create do
+
+  describe '.call' do
+    context 'when all parameters are provided' do
+      it 'creates a product with a nested supplier' do
```
Comment (stevo) [inline spec/graphql/functions/create_spec.rb]: separate case

File: app/graphql/services/nested_attributes.rb [hunk 20]
```diff
@@ -1,10 +1,9 @@
...
   def self.call(model, attributes)
-    nested_attributes = {}
     nested_keys = attributes.keys & model.nested_attributes_options.keys.map(&:to_s)
-    nested_keys.each do |nested_key|
-      nested_attributes["#{nested_key}_attributes"] = attributes.delete(nested_key)
-    end
-    attributes.merge(nested_attributes)
+    return attributes if nested_keys.blank?
+
+    nested_keys.each { |key| attributes["#{key}_attributes"] = attributes.delete(key) }
```
Comment (stevo) [inline app/graphql/services/nested_attributes.rb]: consider using `transform_keys`

**Comment by stevo** [comment]
https://github.com/Selleo/rspec-controller-matchers/blob/master/spec/support/controller_specs_boilerplate.rb
https://github.com/Selleo/rspec-controller-matchers/blob/master/spec/schema.rb

File: Gemfile [hunk 21]
```diff
@@ -54,5 +54,13 @@ group :development do
   gem 'graphiql-rails'
 end
 
+group :test do
+  gem 'factory_girl_rails'
```
Comment (tb) [inline Gemfile]: we use it for seeds so its in global scope, we dont need it here again

File: spec/graphql/functions/create_spec.rb [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+                'reorder_level' => 6056,
+                'target_level' => 5822,
+                'discontinued' => false,
+                'minimum_reorder_quantity' => 7040,
+                'category' => 'Electronics',
+                'supplier' =>
+                  {
+                    'name' => 'Supplier #1',
+                    'webpage' => 'http://example.com',
+                    'notes' => 'Something note'
```
Comment (tb) [inline spec/graphql/functions/create_spec.rb]: In line 57+ we have `product_code: '6393283882',` etc. Cant we define fields/values once and use it here with stringify keys and without it in expect. After all it should be the same.

This applies to all those tests.

**Review by tb — CHANGES_REQUESTED** [review]

File: app/graphql/functions/create.rb [hunk 23]
```diff
@@ -1,14 +1,16 @@
...
   end
 
   def call(obj, args, ctx)
-    attributes = args[@param_key].to_h
-    @model.create(Services::NestedAttributes.call(@model, attributes))
+    attributes = args[param_key].to_h
+    model.create(Services::NestedAttributes.call(model, attributes))
   end
+
+  private
```
Comment (ghost) [inline app/graphql/functions/create.rb:13]: Why `private attr_reader` ?

**Pull request closed without merging**